[Migrate] From `WellSQL` to `Room`

ParaskP7 opened the pull request:
Closes: [AINFRA-750](https://linear.app/a8c/issue/AINFRA-750)
Closes: #12 (see ef2bdcf0cca59bf0cc0b15d48866666d6f48b0fa)

---

### Description

This is the final migration PR which replaces [WellSQL](https://github.com/wordpress-mobile/wellsql) with [Room](https://developer.android.com/jetpack/androidx/releases/room). Main changes include but not limited to:

- Replacing `WellSQL` with `Room`. Also, removing every `wordpress` related dependency and repository from the library.
- Replacing `dateCreated` with a new synthetic primary key (`id`).

---

### Testing Steps

Using the [DOAndroid](https://github.com/bloom/DayOne-Android) or/and [PCAndroid](https://github.com/Automattic/pocket-casts-android) project, update `automattic-encryptedlogging` to `20-ef2bdcf0cca59bf0cc0b15d48866666d6f48b0fa` (version which is based on this PR's [latest](https://github.com/Automattic/EncryptedLogging/pull/20/commits/ef2bdcf0cca59bf0cc0b15d48866666d6f48b0fa) commit) and make sure that encrypted logging continues to work as expected:

1. Using `App Inspection` from AS find the `encrypted-logging.db` and its sole `EncryptedLogModel` table. Notice that the table is empty, there are no log to be sent.
2. Make the app crash. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could navigate to `Settings` -> `Developer` and then click `Crash The App`.
3. You might now notice a new db entry within `App Inspection` from AS on that `DETACHED` app (or not, it depends). Now, switch to the `ATTACHED` app and notice the db empty. This is because this db entry/log has been already sent to Sentry already, on app start and after the app crashed. To really notice the entry, trying crashing the app again, this time put the device on offline mode first. Doing that you'll be able to see the entry on the db, even when switching to the newly created `ATTACHED` app.
4. Open [Sentry](https://a8c.sentry.io) and navigate to `Issues`. Find the project you're testing for (for example `day-one-android` or/and `pocket-casts-android`) and check the latest `debug` exceptions. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could search for `Developer caused a crash!` and check the latest crash. Now navigate at the bottom and check the `uuid`. Verify it is the same `UUID` that you saw on that db entry before.
5. Open [Read Encrypted Logs](https://mc.a8c.com/encrypted-logs.php), copy-paste that `UUID` into `Log UUID`, click `View` and make sure you can read the newly uploaded and now decrypted log.

---

Just because of this [dateCreated-to-id](https://github.com/Automattic/EncryptedLogging/pull/20/commits/ef2bdcf0cca59bf0cc0b15d48866666d6f48b0fa) change, you might also want to test the changes introduced by queueing more than one encrypted log and see it working in action:

For your convenience, I did exactly that, see below:

#### 2 Queue Encrypted Logs - With Different Upload States

![db - 2 events - different states](https://github.com/user-attachments/assets/6855b479-10a3-4444-a4b8-6be8ea1ba33f)

![logcat - 2 events - different states](https://github.com/user-attachments/assets/6b9298df-b11c-4aa3-83e8-8caf2ed998d6)

From the Logcat output, you'll notice that the ordered worked as expected. First the encrypted log with upload state equal to 1 (`QUEUED`) was sent. Then, the encrypted log with upload state equal to 3 (`FAILED`) was sent.

#### 3 Queue Encrypted Logs - With Same Upload State

![db - 3 events - same state](https://github.com/user-attachments/assets/03d92359-3439-46fb-ad7e-fef4642d05f5)

![logcat - 3 events - same state](https://github.com/user-attachments/assets/059cc652-5c50-49f0-8011-523d965219d2)

From the Logcat output, you'll notice that the ordered worked as expected. First the encrypted log with id equal to 4 was sent. Then, the encrypted log with id equal to 5 was sent. Finally, the encrypted log with id equal to 6 was sent.

---

Additionally, run this [ReleaseStack_EncryptedLogTest](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/androidTest/kotlin/com/automattic/encryptedlogging/release/ReleaseStack_EncryptedLogTest.kt#L43)  instrumented test locally, and make sure that it runs successfully, and as expected. Note that:

1. It is quite hard to work with this test because `TooManyRequests` can be hit quite easily. At which point you need to wait for a few minutes (15' minutes to half an hour) before being able to verify it is working as expected.
2. In order to verify that this test works as expected one would need to use the debugger and navigate through the flow manually. Otherwise the test might success with a false positive, but actually silently failing on (at least) this `TooManyRequests` problem mentioned above.

ParaskP7 committed:
Cleanup: Remove unused imports from automattic encrypted logging

ParaskP7 committed:
Refactor: Improve imports for encrypted log

ParaskP7 committed:
Refactor: Improve imports for encrypted log store

ParaskP7 committed:
Refactor: Improve imports for encrypted log rest client

ParaskP7 committed:
Refactor: Reformat on encrypted log uploaded into multiple-line

FYI: This line is longer than the normal maximum line length used (120).

ParaskP7 committed:
Refactor: Reformat upload encrypted log error unknown into multiple-line

FYI: This line is longer than the normal maximum line length used (120).

ParaskP7 committed:
Refactor: Improve imports for encrypted log instrumented test

ParaskP7 committed:
Migrate: Replace wellsql with room (skeleton)

FYI: This is a skeleton implementation which mainly removes the
'wordpress-wellsql' library, replacing it with the most
basic 'androidx-room' setup for the project to compile, with tests, both
unit and UI still failing.

PS: 'kapt' got replaced with 'ksp' as part of this change too.

ParaskP7 committed:
Migrate: Update encrypted log entity (to its complete schema)

In addition to making 'EncryptedLogModel' a data class, this change
removes the 'id' column and make 'uuid' the primary key for the
'EncryptedLogEntity' table.

ParaskP7 committed:
Migrate: Add encrypted log dao (and replace encrypted log sql utils)

ParaskP7 committed:
Refactor: Extract get encrypted log for upload private function

FYI: This is done in order to match the previous signature and make the
parent function easier to read.

ParaskP7 committed:
Refactor: Extract get uploading encrypted logs and map to failed func

FYI: This is done in order to make the parent function easier to read.

ParaskP7 committed:
Refactor: Extract upsert/delete encrypted log dao helper functions

FYI: This is done in order to make the store functions easier to read.

ParaskP7 committed:
Test: Extract create test encrypted log model private function

FYI: This is done in order to make the dao test functions easier to
read, reason and deal with.

ParaskP7 committed:
Migration: Replace date created with synthetic primary key

FYI: This is done in order to get rid 'DateTimeUtils' and everything
related to handling 'dateCreated' just for ordering purposes. Using a
synthetic auto-generated primary key is enough and will achieve the same
ordering of encrypted logs, but this time, without depending on the date
that encrypted log was created.

PR Discussions:
- https://github.com/Automattic/EncryptedLogging/pull/17
#discussion_r2120752334
- https://github.com/Automattic/EncryptedLogging/pull/18
#discussion_r2149442437

Related Issue: Consider removing EncryptedLog#dateCreated and
EncryptedLogModel#dateCreated fields #12
- https://github.com/Automattic/EncryptedLogging/issues/12

ParaskP7 commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDao.kt:
```diff
@@ -0,0 +1,66 @@
+
+    /* HELPER FUNCTIONS */
+
+    internal suspend fun insertEncryptedLog(encryptedLog: EncryptedLog) =
```
Nah, I am good with it Copi, thanks for the nitpick though! 🙏

ParaskP7 commented:
👋 @wzieba and about [this](https://github.com/Automattic/EncryptedLogging/pull/17#discussion_r2156627738), I was wondering how to best proceed with it, and whether I should do this change as part of this PR, or not, let discuss. 🙏

1. My main question is not whether we should delete this public [observeEncryptedLogsUploadResult()](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/AutomatticEncryptedLogging.kt#L88-L90) method, but whether we want to also delete the [uploadState](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt#L76-L77). 🤔 FYI: Note if we delete [uploadState](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt#L76-L77) this will mean that we want be able to work effectively with [ReleaseStack_EncryptedLogTest](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/androidTest/kotlin/com/automattic/encryptedlogging/release/ReleaseStack_EncryptedLogTest.kt#L43C16-L43C45).
2. A secondary question is where to put these logs, and what's the best way to deal with that. Just below `_uploadState.value`, where it is updated within [EncryptedLogStore](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt#L46C16-L46C33), or somewhere else? Maybe the library should observe this [observeEncryptedLogsUploadResult()](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/AutomatticEncryptedLogging.kt#L88-L90) method itself, after we make it `internal`. 🤔

Let me know your thoughts and thanks! 🙏

wzieba commented:
> FYI: Note if we delete [uploadState](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt#L76-L77) this will mean that we want be able to work effectively with [ReleaseStack_EncryptedLogTest](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/androidTest/kotlin/com/automattic/encryptedlogging/release/ReleaseStack_EncryptedLogTest.kt#L43C16-L43C45).

As this is e2e test, we could validate if the logs are actually uploaded by querying them, the same like you do manually when testing PRs (this requires proxy though). Alternatively, you can leave `uploadState` as test-only fields and it's more-or-less fine as long as it's `internal`.

> A secondary question is where to put these logs

I'll leave it up to you, I don't think it's that important. `handleFailedUpload` looks like a good candidate.

> Maybe the library should observe this [observeEncryptedLogsUploadResult()](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/AutomatticEncryptedLogging.kt#L88-L90) method itself

I don't think it's necessary - it'll be better to log in the emitter.

wzieba commented on encryptedlogging/src/androidTest/kotlin/com/automattic/encryptedlogging/release/ReleaseStack_EncryptedLogTest.kt:
```diff
@@ -61,23 +52,23 @@ internal class ReleaseStack_EncryptedLogTest {
         encryptedLogStore.uploadState.value = null
         cleanSharedPreferencesState()
-        WellSql.delete(EncryptedLogModel::class.java).execute()
+        database.encryptedLogDao.deleteEncryptedLogs()
```
Minor, but I think we could close db here as well

```suggestion
        database.encryptedLogDao.deleteEncryptedLogs()
        database.close()
```

ParaskP7 replied:
Thanks for this comment @wzieba , I actually tried that but it doesn't work because when the database is closed (after the first test, which succeeds), then that makes any subsequent test(s) fail. This is different to the in-memory database on unit tests... 🤷

wzieba replied:
ah I see 👍 `database` here is from Singleton, which survives the JUnit singular execution. I guess it's not a big deal, but I wonder what would happen if you'd cancel the test execution in the middle? Will `@After` method be called and will DB be cleared?

ParaskP7 replied:
> ...what would happen if you'd cancel the test execution in the middle?

How and why would you do that @wzieba? 🤔

> Will @After method be called and will DB be cleared?

When you cancel the execution, even if the `@After` method isn't called, and the database isn't cleared, how would that be a problem to us? 🤔

wzieba replied:
> How and why would you do that @wzieba? 🤔

Just stop the process, Gradle crash, disconnect the phone - many possibilities.

> how would that be a problem to us? 🤔

The database will contain events from previous executions, which might affect current test run.

wzieba commented on encryptedlogging/src/androidTest/kotlin/com/automattic/encryptedlogging/release/ReleaseStack_EncryptedLogTest.kt:
```diff
@@ -159,7 +150,7 @@ internal class ReleaseStack_EncryptedLogTest {
     private fun cleanSharedPreferencesState() {
         preferenceUtils.getPreferences().edit().putLong(
-            ENCRYPTED_LOG_UPLOAD_UNAVAILABLE_UNTIL_DATE,
+            com.automattic.encryptedlogging.store.ENCRYPTED_LOG_UPLOAD_UNAVAILABLE_UNTIL_DATE,
```
This and similar imports above can be reduced for readability.

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/model/encryptedlogging/EncryptedLog.kt:
```diff
@@ -1,80 +1,69 @@
-        this.id = id
-    }
-
+@Entity(
+    tableName = "EncryptedLogEntity",
+)
+internal data class EncryptedLogModel(
```
Minor again, but I think we can simply rename `EncryptedLogModel` - I don't do this in Woo during migration as diffs would be huge. 

```suggestion
internal data class EncryptedLogEntity(
```

ParaskP7 replied:
FYI: This is now done: ad976f6fff38fd1f2dc7560cab60ab04f86b1078

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/network/rest/wpcom/encryptedlog/EncryptedLogRestClient.kt:
```diff
@@ -23,9 +21,9 @@ internal class EncryptedLogRestClient(
+                cont.resume(UploadEncryptedLogResult.LogUploaded)
             }, { error ->
-                cont.resume(LogUploadFailed(mapError(error)))
+                cont.resume(UploadEncryptedLogResult.LogUploadFailed(mapError(error)))
```
Here, and in multiple other places, I think we should import members

ParaskP7 replied:
You mean you preferred it the way it was @wzieba, `LogUploadFailed(mapError(error))` versus `UploadEncryptedLogResult.LogUploadFailed(mapError(error))`? 🤔

wzieba replied:
Yes - I think there's no value introducing a qualified name

ParaskP7 replied:
Thanks @wzieba, the value of introducing a qualified name, IMHO, is that you get by with at least one less import:

Instead of:

```kotlin
import com.automattic.encryptedlogging.network.rest.wpcom.encryptedlog.UploadEncryptedLogResult.LogUploaded
import com.automattic.encryptedlogging.network.rest.wpcom.encryptedlog.UploadEncryptedLogResult.LogUploadFailed
```

or a `*`, which should be avoided:

```kotlin
import com.automattic.encryptedlogging.network.rest.wpcom.encryptedlog.UploadEncryptedLogResult.*
```

You now have:

```kotlin
import com.automattic.encryptedlogging.network.rest.wpcom.encryptedlog.UploadEncryptedLogResult
```

---

Plus, for me personally, and I know this is a personal thing, but I rather read `UploadEncryptedLogResult.LogUploaded` than `LogUploaded`, that's why I did this change. I understand you prefer it the other way around.

FYI: Same logic applies to all the import changes I did.

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDao.kt:
```diff
@@ -0,0 +1,66 @@
+
+@Dao
+internal abstract class EncryptedLogDao {
+    @Query("""
+        SELECT * FROM EncryptedLogEntity 
+        WHERE uuid = :uuid
+    """)
+    internal abstract suspend fun getEncryptedLog(uuid: String): EncryptedLogModel?
```
This method is used only in tests, please consider removing it
```suggestion
    
```

ParaskP7 replied:
Yea, this and `deleteEncryptedLogs()` are tests only, I too was thinking about it @wzieba. But, if I go removing them, how would I go about the DAO unit test, as well as that UI test which depends on deleting all encrypted logs. 🤔

---

FYI: I thought about using `getEncryptedLog(uploadStates)` instead, but refrain from doing that, thinking that it's okay for us to expose a test-only DAO method for this library, that is, rather than try and find a test workaround like:

Instead of:
```kotlin
    fun `xzy`() = runTest {
        sut.getEncryptedLog(TEST_UUID)
    }
```

Now do:
```kotlin
 @Test
    fun `xzy`() = runTest {
        getEncryptedLog()
    }

    private suspend fun getEncryptedLog(): EncryptedLogModel? {
        val uploadStates = listOf(
            EncryptedLogUploadState.QUEUED,
            EncryptedLogUploadState.UPLOADING,
            EncryptedLogUploadState.FAILED
        ).map { it.value }
        return sut.getEncryptedLog(uploadStates)
    }
```

And even doing that, we'll still have `test insert multiple encrypted logs` failing because it is based on uuid. 🤷

What's your recommendation here?

FYI: I am honestly fine with having both those test-only DAO methods to helps testing things easier, maybe it will be enough to use a `@TestOnly` annotation, wdyt? 🤔

wzieba replied:
Having test-only methods in the code is an antipattern. I understand we can argue of having it in the case of e2e test, which is unconventional in our case, but for a simple unit test we should not need it.

If we do, this means that we don't actually test the contract we established with the API.

So e.g. let's take the first test

```kotlin
  @Test
    fun `test insert encrypted log`() = runTest {
        // Assert that there are no encrypted logs with the test uuid
        assertThat(sut.getEncryptedLog(TEST_UUID)).isNull() // <- this assertion tests the test (check if the db is empty at the beginning of the test) and not actual implementation - we should remove it

        // Insert an encrypted log with uuid
        val logToBeInserted = createTestEncryptedLogModel()
        sut.insertEncryptedLog(logToBeInserted)

        // Assert that the encrypted log from the DB is the same as the one we inserted (ignoring id)
        assertThat(sut.getEncryptedLog(TEST_UUID)) // <- this assertion tests imaginery case: we never query logs by their UUID in the code, so why test it?
            .usingRecursiveComparison()
            .ignoringFields("id")
            .isEqualTo(logToBeInserted)
    }
```

my proposal is something like this

```kotlin
    @Test
    fun `test insert encrypted log`() = runTest {
        val logToBeInserted = createTestEncryptedLogModel()

        sut.insertEncryptedLog(logToBeInserted)

        assertThat(sut.getEncryptedLogs(QUEUED.value))
            .hasSize(1)
            .first()
            .usingRecursiveComparison()
            .ignoringFields("id")
            .isEqualTo(logToBeInserted)
    }

```

If you want, you can be more explicit that `logToBeInserted` has `QUEUED` state by passing the state in `createTestEncryptedLogModel`.

ParaskP7 replied:
Thanks for the reply @wzieba ! 🙏 

> Having test-only methods in the code is an antipattern.

👍

> I understand we can argue of having it in the case of e2e test, which is unconventional in our case...

> ...but for a simple unit test we should not need it.
> If we do, this means that we don't actually test the contract we established with the API.

👍 

> my proposal is something like this

Thanks for this proposal, I'll proceed with this change and remove the unit test-only DAO function, keeping only the UI test-only DAO function. 👍

FYI: I am still not sure what to do with the `test insert multiple encrypted logs` test, worst-case scenario, I'll remove it. 🤷

ParaskP7 replied:
FYI: This is now done: f264979d6349d7086eac5956a782621699517d23

wzieba replied:
Thank you!

ParaskP7 replied:
Thank YOU! 🙇

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDao.kt:
```diff
@@ -0,0 +1,66 @@
+    internal abstract suspend fun getEncryptedLogsCount(uploadState: Int): Int
+
+    @Insert
+    internal abstract suspend fun insertEncryptedLog(encryptedLog: EncryptedLogModel)
```
This is only used in this DAO, let's make it not accessible outside if it's not necessary 
```suggestion
    protected abstract suspend fun insertEncryptedLog(encryptedLog: EncryptedLogModel)
```

ParaskP7 replied:
Thanks @wzieba, this is only used in this DAO because of the helper functions ([related](https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2163819958)), plus used in tests, I would rather keep this as `internal` for now. 🤔

PS: [Same](https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2163814097)

ParaskP7 replied:
Related: https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2164244799

wzieba replied:
Sure - as helper functions are removed and this method is used in Store, then we can't make it dao-only.

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDao.kt:
```diff
@@ -0,0 +1,66 @@
+    internal abstract suspend fun upsertEncryptedLogs(encryptedLogs: List<EncryptedLogModel>)
+
+    @Delete
+    internal abstract suspend fun deleteEncryptedLog(encryptedLog: EncryptedLogModel)
```
Same as above

```suggestion
    protected abstract suspend fun deleteEncryptedLog(encryptedLog: EncryptedLogModel)
```

ParaskP7 replied:
PS: [Similar](https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2164000565)

ParaskP7 replied:
Related: https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2164244799

wzieba replied:
https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2166129829

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDao.kt:
```diff
@@ -0,0 +1,66 @@
+import com.automattic.encryptedlogging.model.encryptedlogging.EncryptedLogModel
+
+@Dao
+internal abstract class EncryptedLogDao {
```
💡 I don't love the ambiguity of this class: some of its parameters accept domain objects and some of them database entities. I think it'd be nicer to remove code below "/* HELPER FUNCTIONS */" and introduce either a mapper object or extension functions to map from domain. Then make `EncryptedLogDao` accept only DB entities.

ParaskP7 replied:
Yea, I wasn't too sure about that too @wzieba, but wanted to keep the changes to a minimum without introducing to many classes, like mappers. But, I do like this rule of keeping the DAO straight, not having it accept anything else than database entries. I'll think of a better way... 🤔

ParaskP7 replied:
This is now done: 28502efa3064ed0019d3e384d6dc3f0364e014b0

wzieba replied:
Thanks!

ParaskP7 commented:
Thanks for the reply @wzieba ! 🙇

> ...and whether I should do this change as part of this PR

What about this, should I do this change as part of this PR, or better create another one right after, wdyt? 🤔

---

> As this is e2e test, we could validate if the logs are actually uploaded by querying them, the same like you do manually when testing PRs (this requires proxy though).

🤔

> Alternatively, you can leave uploadState as test-only fields and it's more-or-less fine as long as it's internal.

I'll go with leaving `uploadState` as a test-only field for now, thanks! 👍 

FYI: It doesn't seem worthy to me atm to change the e2e test login by querying them. 😊

> I'll leave it up to you, I don't think it's that important. handleFailedUpload looks like a good candidate.

🤔 🗒️

> I don't think it's necessary - it'll be better to log in the emitter.

👍

wzieba commented on encryptedlogging/src/test/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDaoTest.kt:
```diff
@@ -0,0 +1,236 @@
+
+        // Assert that the encrypted log in the DB is the one with the correct upload state
+        val updatedLogFromDB = sut.getEncryptedLog(TEST_UUID)
+        assertThat(requireNotNull(updatedLogFromDB?.uploadState)).isEqualTo(newUploadState)
+        // This verifies the expected state as well but separating the initial assertion is valuable to show intent
```
I understand the intention, but IMO it's not needed. In the second assertion, the different `uploadState` will be highlighted. My suggestion would be to keep tests as simple and focused as possible

```suggestion

```

ParaskP7 replied:
Yea, I agree @wzieba, I just converted this test with minimal changes to keep it intact.

PS: [Same](https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2163837693)

ParaskP7 replied:
FYI: This is now done: 09d587eec492d6c8ad5a57a69f2e8d90e56cfadb

wzieba commented on encryptedlogging/src/test/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDaoTest.kt:
```diff
@@ -0,0 +1,236 @@
+
+    @Test
+    fun `test uploading encrypted logs for random number`() = runTest {
+        Random.nextInt(100).let { numberOfLogs ->
```
We should not use random in tests. I don't think this brings any value, comparing to a predefined constant, while introducing randomness which is difficult to debug.

ParaskP7 replied:
PS: [Similar](https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2164016781)

ParaskP7 replied:
FYI: This is now done: 4b1e4fa85ee71dbf7a7d86de4a9c6ac45bed9419

wzieba reviewed (commented):
Nice work @ParaskP7 ! I left some comments. It'd be great to revert changes related to not importing members, to make the diff of this PR smaller.

ParaskP7 commented:
> It'd be great to revert changes related to not importing members...

About this @wzieba, I really wouldn't like to revert this change and still think (at least most of them) are beneficial. 😊

> ...to make the diff of this PR smaller.

I hear this argument, but this way I would never introduce any refactor change on any PR, that is, unless I create a PR that will be dedicated to refactor, which we both know we never go do afterwords. Thus, I usually add such changes, as separate commits, to be reviewed in isolation to the main commits. Those commits are prefaced with `Refactor` in the commit title so that the reviewer could ignore them at will and focus on the main `Feature` (or in this case `Migration`) commits. 💭

Now, I understand that you might be reviewing this PR in GitHub and thus it doesn't help when you see the whole list of changes from the main `Files changed` tab. But, what if you check the changes one commit at a time, or use the IDE (AS) to review this PR instead? 😊

---

Btw, this whole discussion will help me understand how to work with you on subsequent PRs from here onwards, so I think it is worth discussing it, and get it out of the way. This way we'll avoid any future such behaviors. 🙏

PS: I still love the `Refactor` commits I usually add here-and-there... 😊

ParaskP7 committed:
Refactor: Rename encrypted log model to entity

PR Comment:  https://github.com/Automattic/EncryptedLogging/pull/20
#discussion_r2163784033

ParaskP7 committed:
Test: Remove test-only get encrypted log dao function

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/20
#discussion_r2163803345

ParaskP7 committed:
Refactor: Remove helper function from dao (use directly instead)

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/20
#discussion_r2163819958

ParaskP7 committed:
Test: Remove unnecessary dao test assertion

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/20
#discussion_r2163833639

ParaskP7 committed:
Test: Replace random with concrete number of logs for doa test

FYI: 10 number of logs is enough.

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/20
#discussion_r2163837693

ParaskP7 commented:
Thanks for your thorough review on this @wzieba, I applied almost all of your comment (minus the `import` related ones, plus [this](https://github.com/Automattic/EncryptedLogging/pull/20#discussion_r2163868567)). Could you please take another look. ❤️ 🙇 

---

Also, about [this](https://github.com/Automattic/EncryptedLogging/pull/20#issuecomment-3000123224) question below, please let me know how to proceed. 🙇

> ...and whether I should do this change as part of this PR

What about this, should I do this change as part of this PR, or better create another one right after, wdyt? 🤔

wzieba committed:
Migration: Introduce type converter for upload state

h/t to @wzieba - PR Comment: https://github.com/Automattic/
EncryptedLogging/pull/20#pullrequestreview-2957158525

wzieba commented on encryptedlogging/src/test/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDaoTest.kt:
```diff
@@ -0,0 +1,232 @@
+            .isEqualTo(numberOfLogs.toLong())
+    }
+
+    @Test
+    fun `test get encrypted logs for upload includes QUEUED logs`() = runTest {
+        sut.insertEncryptedLog(createTestEncryptedLogEntity(uploadState = EncryptedLogUploadState.QUEUED))
+
+        val uploadStates = listOf(
+            EncryptedLogUploadState.QUEUED,
+            EncryptedLogUploadState.FAILED
+        ).map { it.value }
+        assertThat(sut.getEncryptedLog(uploadStates)).isNotNull
+    }
+
+    @Test
+    fun `test get encrypted logs for upload includes FAILED logs`() = runTest {
+        sut.insertEncryptedLog(createTestEncryptedLogEntity(uploadState = EncryptedLogUploadState.FAILED))
+
+        val uploadStates = listOf(
+            EncryptedLogUploadState.QUEUED,
+            EncryptedLogUploadState.FAILED
+        ).map { it.value }
+        assertThat(sut.getEncryptedLog(uploadStates)).isNotNull
+    }
+
+    @Test
+    fun `test get encrypted logs for upload does not include UPLOADING logs`() = runTest {
+        sut.insertEncryptedLog(createTestEncryptedLogEntity(uploadState = EncryptedLogUploadState.UPLOADING))
+
+        val uploadStates = listOf(
+            EncryptedLogUploadState.QUEUED,
+            EncryptedLogUploadState.FAILED
+        ).map { it.value }
+        assertThat(sut.getEncryptedLog(uploadStates)).isNull()
+    }
```
I think these 3 tests could be either reduced to a single test or 2 of them removed: these states are not different in any way, there's no different logic for them, IMO they're quite redundant.

ParaskP7 replied:
Right, I am now actually thinking of removing all 3 of them, and only keep that 4th one, which is about ordering, which also includes the same testing that is happening on all those 3 (to be removed) tests above. Wdyt @wzieba? 🤔

wzieba replied:
Sounds good!

ParaskP7 replied:
FYI: This is now done: 262d4a9f44cfd222b3664278d4f1ed083bbab784

wzieba reviewed (commented):
Thank you for applying suggestions. During the review, I noticed a placed for improvement to use TypeConverter instead of constant `.value`. I think it simplifies the solution. Here's the idea, if you'd like you could cherry pick it 6e6d8d4ffac52eb87ea675329a8f33c6589cf9ff

wzieba commented:
> ...and whether I should do this change as part of this PR

I'll leave it up to you, for me personally it doesn't matter in which PR the change will be applied 👍

ParaskP7 commented:
👋 @wzieba and thanks for taking another look at that! 🙏 

> During the review, I noticed a placed for improvement to use TypeConverter instead of constant .value. I think it simplifies the solution. Here's the idea, if you'd like you could cherry pick it https://github.com/Automattic/EncryptedLogging/commit/6e6d8d4ffac52eb87ea675329a8f33c6589cf9ff

Interesting, let me check and possibly apply this change, like it, thanks for sharing! 🥇

> I'll leave it up to you, for me personally it doesn't matter in which PR the change will be applied 👍

Okay, then I'll apply this change here just to get this out of the way as well, thanks! 👍

ParaskP7 committed:
Test: Remove unnecessary duplicate dao tests for upload states

Keeping this `test get encrypted logs for upload is in correct order`
only should be enough to cover testing 'getEncryptedLog(uploadStates)'.

ParaskP7 committed:
Sdk: Remove observe encrypted logs upload result function from sdk

Clients used this functionality to observe the upload result of
encrypted logs. However, after speaking to multiple clients, this
functionality was mostly used for logging purposes, something the
library could/will do itself, without having to delegate it to clients.

Only one client (WCAndroid) is currently using this functionality to
also track encrypted logs upload events. However, after reaching out to
this team, they informed us they are actually not using those tracks too
much (if at all). Thus, it is safe to assume that no clients actually
depend on this functionality and that it can be effectively removed from
the sdk.

Discussion: https://github.com/Automattic/EncryptedLogging/pull/17
#discussion_r2145393073
PR Comment:https://github.com/Automattic/EncryptedLogging/pull/20
#issuecomment-3000059069

ParaskP7 commented:
👋 @wzieba, one last review! 🙏

> Interesting, let me check and possibly apply this change, like it, thanks for sharing! 🥇

FYI: This is now done: 1be9273c14f48c30c2ed0761ba986a0cc141f5dd (many thanks for this suggestions, so much better) ❤️

> Okay, then I'll apply this change here just to get this out of the way as well, thanks! 👍

FYI: This is now done: 3c19605c132a5c19b50fc58ca1bb54639a534aa1

wzieba reviewed (approved):
Looks great, thanks for working on this and suggestions! 🙌

ParaskP7 commented:
Awesome @wzieba , thank YOU for ALL the suggestions, you ROCK! 🙇 ❤️ 🚀
